README
NAME
    PYX::XMLSchema::List - Processing PYX data or file and print list of XML
    schemas.

SYNOPSIS
     use PYX::XMLSchema::List;

     my $obj = PYX::XMLSchema::List->new(%parameters);
     $obj->parse($pyx, $out);
     $obj->parse_file($input_file, $out);
     $obj->parse_handler($input_file_handler, $out);
     $obj->reset;
     my $stats_hr = $obj->stats;

METHODS
  "new"
     my $obj = PYX::XMLSchema::List->new(%parameters);

    Constructor.

    *       "output_handler"

            Output handler.

            Default value is \*STDOUT.

    Returns instance of object.

  "parse"
     $obj->parse($pyx, $out);

    Parse PYX text or array of PYX text and print list of XML schemas of PYX
    input. If $out not present, use 'output_handler'.

    Returns undef.

  "parse_file"
     $obj->parse_file($input_file, $out);

    Parse file with PYX data and print list of XML schemas of PYX input. If
    $out not present, use 'output_handler'.

    Returns undef.

  "parse_handler"
     $obj->parse_handler($input_file_handler, $out);

    Parse PYX handler and print list of XML schemas of PYX input. If $out
    not present, use 'output_handler'.

    Returns undef.

  "reset"
     $obj->reset;

    Resets internal structure with statistics.

    Returns undef.

  "stats"
     my $stats_hr = $obj->stats;

    Gets statistics structure.

    Returns undef.

ERRORS
     new():
             From Class::Utils::set_params():
                     Unknown parameter '%s'.

EXAMPLE1
     use strict;
     use warnings;

     use PYX::XMLSchema::List;

     # Example data.
     my $pyx = <<'END';
     (foo
     Axmlns:bar http://bar.foo
     Axmlns:foo http://foo.bar
     Afoo:bar baz
     (foo:bar
     Axml:lang en
     Abar:foo baz
     )foo:bar
     )foo
     END

     # PYX::XMLSchema::List object.
     my $obj = PYX::XMLSchema::List->new;

     # Parse.
     $obj->parse($pyx);

     # Output:
     # [ bar ] (E: 0000, A: 0001) http://bar.foo
     # [ foo ] (E: 0001, A: 0001) http://foo.bar
     # [ xml ] (E: 0000, A: 0001)

DEPENDENCIES
    Class::Utils, Error::Pure, List::Util, PYX::Parser, Readonly.

SEE ALSO
    PYX A perl module for PYX handling.

    Task::PYX
        Install the PYX modules.

REPOSITORY
    <https://github.com/michal-josef-spacek/PYX-XMLSchema-List>

AUTHOR
    Michal Josef Špaček <mailto:skim@cpan.org>

    <http://skim.cz>

LICENSE AND COPYRIGHT
    © 2015-2023 Michal Josef Špaček

    BSD 2-Clause License

VERSION
    0.07



Changes
0.07

0.06 2023-10-19T10:12:48+02:00
 - Fix Makefile.PL which fails with no '.' in @INC.

0.05 2023-10-19T10:09:18+02:00
 - Improve LICENSE AND COPYRIGHT section in doc.
 - Improve METHODS section in doc.
 - Improve SYNOPSIS section in doc.
 - Move bugtracker to github.
 - Rename example to better name.
 - Update Module::Install to 1.21 version.
 - Update author github username.
 - Update author name.
 - Update copyright years.

0.04 2015-10-29T21:44:46+01:00
 - Fix warning in length(undef).

0.03 2015-10-25T09:23:29+01:00
 - Fix test requires.

0.02 2015-10-22T19:44:43+02:00
 - Add reset() method.
 - Add stats() method.
 - Add tests.
 - Clean and optimalize code.
 - Fix output (trailing whitespace on end of line).
 - Improve tests to better coverity.
 - Improve SEE ALSO section in doc.
 - Move author tests to xt/ directory.
 - Update Module::Install to 1.16 version.

0.01 2015-04-01T01:11:02+02:00
 - First version.


List.pm
package PYX::XMLSchema::List;

use strict;
use warnings;

use Class::Utils qw(set_params);
use Error::Pure qw(err);
use List::Util qw(reduce);
use PYX::Parser;
use Readonly;

# Constants.
Readonly::Scalar our $EMPTY_STR => q{};
Readonly::Scalar our $SPACE => q{ };

our $VERSION = 0.07;

# Constructor.
sub new {
	my ($class, @params) = @_;
	my $self = bless {}, $class;

	# Output handler.
	$self->{'output_handler'} = \*STDOUT;

	# Process params.
	set_params($self, @params);

	# PYX::Parser object.
	$self->{'_pyx_parser'} = PYX::Parser->new(
		'callbacks' => {
			'attribute' => \&_call_attribute,
			'final' => \&_call_final,
			'start_element' => \&_call_start_element,
		},
		'non_parser_options' => {
			'schemas' => {},
		},
		'output_handler' => $self->{'output_handler'},
	);

	# Object.
	return $self;
}

# Parse pyx text or array of pyx text.
sub parse {
	my ($self, $pyx, $out) = @_;
	$self->{'_pyx_parser'}->parse($pyx, $out);
	return;
}

# Parse file with pyx text.
sub parse_file {
	my ($self, $file, $out) = @_;
	$self->{'_pyx_parser'}->parse_file($file, $out);
	return;
}

# Parse from handler.
sub parse_handler {
	my ($self, $input_file_handler, $out) = @_;
	$self->{'_pyx_parser'}->parse_handler($input_file_handler, $out);
	return;
}

# Reset parser.
sub reset {
	my $self = shift;
	$self->{'_pyx_parser'}->{'non_parser_options'}->{'schemas'} = {};
	return;
}

# Gets statistics structure.
sub stats {
	my $self = shift;
	my $schemas_hr = $self->{'_pyx_parser'}->{'non_parser_options'}
		->{'schemas'};
	return $schemas_hr;
}

# Attribute callback.
sub _call_attribute {
	my ($pyx_parser_obj, $key, $val) = @_;
	my $schemas_hr = $pyx_parser_obj->{'non_parser_options'}->{'schemas'};
	if (my ($first, $sec) = _parse_schema($key)) {

		# Get URL for XML schema.
		if ($first eq 'xmlns') {
			my $schema = $sec;
			if (! exists $schemas_hr->{$schema}) {
				_init_struct($schemas_hr, $schema, $val);
			} else {
				$schemas_hr->{$schema}->[0] = $val;
			}

		# Add attribute to XML schema statistics.
		} else {
			my $schema = $first;
			_init_struct($schemas_hr, $schema);
			$schemas_hr->{$schema}->[1]->{'attr'}++;
		}
	}
	return;
}

# Finalize callback.
sub _call_final {
	my $pyx_parser_obj = shift;
	my $schemas_hr = $pyx_parser_obj->{'non_parser_options'}->{'schemas'};
	my $out = $pyx_parser_obj->{'output_handler'};
	my $max_string = reduce { length($a) > length($b) ? $a : $b } keys %{$schemas_hr};
	my $max_len = defined $max_string ? length $max_string : 0;
	foreach my $key (sort keys %{$schemas_hr}) {
		printf {$out} "[ %-${max_len}s ] (E: %04d, A: %04d)%s\n", $key,
			$schemas_hr->{$key}->[1]->{'element'},
			$schemas_hr->{$key}->[1]->{'attr'},
			$schemas_hr->{$key}->[0] ne $EMPTY_STR
				? $SPACE.$schemas_hr->{$key}->[0]
				: $EMPTY_STR;
	}
	if (! keys %{$schemas_hr}) {
		print {$out} "No XML schemas.\n";
	}
	return;
}

# Start of element callback.
sub _call_start_element {
	my ($pyx_parser_obj, $elem) = @_;
	my $schemas_hr = $pyx_parser_obj->{'non_parser_options'}->{'schemas'};
	if (defined(my $schema = _parse_schema($elem))) {
		_init_struct($schemas_hr, $schema);
		$schemas_hr->{$schema}->[1]->{'element'}++;
	}
	return;
}

# Initialize XML schema structure.
sub _init_struct {
	my ($schemas_hr, $schema, $uri) = @_;
	if (! defined $uri) {
		$uri = $EMPTY_STR;
	}
	if (! exists $schemas_hr->{$schema}) {
		$schemas_hr->{$schema} = [$uri, {
			'attr' => 0,
			'element' => 0,
		}];
	}
	return;
}

# Parse XML schema from string.
sub _parse_schema {
	my $string = shift;
	if ($string =~ m/^(.+?):(.+)$/ms) {
		return wantarray ? ($1, $2) : $1;
	}
	return;
}

1;

__END__

=pod

=encoding utf8

=head1 NAME

PYX::XMLSchema::List - Processing PYX data or file and print list of XML schemas.

=head1 SYNOPSIS

 use PYX::XMLSchema::List;

 my $obj = PYX::XMLSchema::List->new(%parameters);
 $obj->parse($pyx, $out);
 $obj->parse_file($input_file, $out);
 $obj->parse_handler($input_file_handler, $out);
 $obj->reset;
 my $stats_hr = $obj->stats;

=head1 METHODS

=head2 C<new>

 my $obj = PYX::XMLSchema::List->new(%parameters);

Constructor.

=over 8

=item * C<output_handler>

Output handler.

Default value is \*STDOUT.

=back

Returns instance of object.

=head2 C<parse>

 $obj->parse($pyx, $out);

Parse PYX text or array of PYX text and print list of XML schemas of PYX input.
If C<$out> not present, use 'output_handler'.

Returns undef.

=head2 C<parse_file>

 $obj->parse_file($input_file, $out);

Parse file with PYX data and print list of XML schemas of PYX input.
If C<$out> not present, use 'output_handler'.

Returns undef.

=head2 C<parse_handler>

 $obj->parse_handler($input_file_handler, $out);

Parse PYX handler and print list of XML schemas of PYX input.
If C<$out> not present, use 'output_handler'.

Returns undef.

=head2 C<reset>

 $obj->reset;

Resets internal structure with statistics.

Returns undef.

=head2 C<stats>

 my $stats_hr = $obj->stats;

Gets statistics structure.

Returns undef.

=head1 ERRORS

 new():
         From Class::Utils::set_params():
                 Unknown parameter '%s'.

=head1 EXAMPLE1

=for comment filename=list_xml_schemas.pl

 use strict;
 use warnings;

 use PYX::XMLSchema::List;

 # Example data.
 my $pyx = <<'END';
 (foo
 Axmlns:bar http://bar.foo
 Axmlns:foo http://foo.bar
 Afoo:bar baz
 (foo:bar
 Axml:lang en
 Abar:foo baz
 )foo:bar
 )foo
 END

 # PYX::XMLSchema::List object.
 my $obj = PYX::XMLSchema::List->new;

 # Parse.
 $obj->parse($pyx);

 # Output:
 # [ bar ] (E: 0000, A: 0001) http://bar.foo
 # [ foo ] (E: 0001, A: 0001) http://foo.bar
 # [ xml ] (E: 0000, A: 0001)

=head1 DEPENDENCIES

L<Class::Utils>,
L<Error::Pure>,
L<List::Util>,
L<PYX::Parser>,
L<Readonly>.

=head1 SEE ALSO

=over

=item L<PYX>

A perl module for PYX handling.

=item L<Task::PYX>

Install the PYX modules.

=back

=head1 REPOSITORY

L<https://github.com/michal-josef-spacek/PYX-XMLSchema-List>

=head1 AUTHOR

Michal Josef Špaček L<mailto:skim@cpan.org>

L<http://skim.cz>

=head1 LICENSE AND COPYRIGHT

© 2015-2023 Michal Josef Špaček

BSD 2-Clause License

=head1 VERSION

0.07

=cut


Makefile.PL
use lib '.';
use strict;
use warnings;

use inc::Module::Install;

# Definition.
abstract 'Processing PYX data or file and print list of XML schemas.';
author 'Michal Josef Spacek <skim@cpan.org>';
author_requires 'English' => 0;
author_requires 'File::Object' => 0.07;
author_requires 'Test::More' => 0;
author_requires 'Test::NoWarnings' => 0;
author_requires 'Test::Pod' => 0;
author_requires 'Test::Pod::Coverage' => 0;
license 'bsd';
name 'PYX-XMLSchema-List';
readme_from 'List.pm';
recursive_author_tests('xt');
requires 'Class::Utils' => 0.05;
requires 'Error::Pure' => 0.16;
requires 'List::Util' => 0;
requires 'PYX::Parser' => 0.02;
requires 'Readonly' => 0;
requires 'perl' => '5.6.0';
resources 'bugtracker' => 'https://github.com/michal-josef-spacek/PYX-XMLSchema-List/issues';
resources 'homepage' => 'https://github.com/michal-josef-spacek/PYX-XMLSchema-List';
resources 'repository' => 'git://github.com/michal-josef-spacek/PYX-XMLSchema-List';
test_requires 'English' => 0;
test_requires 'File::Object' => 0.07;
test_requires 'Error::Pure::Utils' => 0.22;
test_requires 'Perl6::Slurp' => 0;
test_requires 'Test::More' => 0;
test_requires 'Test::NoWarnings' => 0;
test_requires 'Test::Output' => 1.02;
tests_recursive;
version '0.07';

# Run.
WriteAll();
